src/Mapper.cpp

```diff
@@ -1346,3 +1346,223 @@ map<int, CGRANode*>* Mapper::getReorderPath(map<CGRANode*, int>* t_path) {
+        jsonFile<<"     \"Opt2TileXY\":{"<<endl;
+        int idx = 0;
+        for (DFGNode* dfgNode: t_dfg->nodes) {
+                // write dfgnodeID, mapped CGRANode X and Y coordinates
+                jsonFile<<"             \""<<dfgNode->getID()<<"\": {"<<endl; // opt id
+                jsonFile<<"                     \"x\":"<<m_mapping[dfgNode]->getX()<<","<<endl; // opt mapped tile x coordinate
+                jsonFile<<"                     \"y\":"<<m_mapping[dfgNode]->getY()<<endl; // opt mapped tile y coordinate
+                idx += 1;
+                if(idx < t_dfg->nodes.size()) jsonFile<<"             },"<<endl;
+                else jsonFile<<"        }"<<endl;
+        }
+        jsonFile<<"     },"<<endl;
+
+        jsonFile<<"     \"Tile2Level\":{"<<endl;
+        // generate level informations of current mapping results
+        map<int, vector<CGRANode*>> FanIO_CGRANodes;
+        vector<int> FanIOs;
+        int numTiles = 0;
+        for (int i=0; i<t_cgra->getRows(); ++i) {
+                for (int j=0; j<t_cgra->getColumns(); ++j) {
+
+                        // record the number of FanIO for each tile. 
+                        CGRANode* currentCGRANode = t_cgra->nodes[i][j];
+                        if(currentCGRANode->getInLinks() == 0) continue; // only record tiles that have DFGNodes mapped
+                        int FanIO = max(currentCGRANode->getInLinks()->size(), currentCGRANode->getOutLinks()->size());
+                        FanIO_CGRANodes[FanIO].push_back(currentCGRANode);
+
+                        // record FanIO in the list if it appears for the first time
+                        if (find(FanIOs.begin(), FanIOs.end(), FanIO) == FanIOs.end()){
+                                FanIOs.push_back(FanIO);
+                        }
+
+                        numTiles++;
+                }
+        }
+        std::sort(FanIOs.rbegin(), FanIOs.rend()); // sort FanIOs from big to small
+        idx = 0;
+        for(int level = 0; level < FanIOs.size(); level++){
+                int FanIO = FanIOs[level];
+                vector<CGRANode*> tiles = FanIO_CGRANodes[FanIO];
+                for(auto tile : tiles){
+                        idx += 1;
+                        if(idx < numTiles) jsonFile<<"          \""<<tile->getID()<<"\":"<<level<<","<<endl;
+                        else jsonFile<<"             \""<<tile->getID()<<"\":"<<level<<endl;
+                }
+        }
+        jsonFile<<"     }"<<endl;
+
+        jsonFile<<"}"<<endl;
+}
+
+// read from the referenced mapping results json file and generates variables for incremental mapping
+int Mapper::readRefMapRes(CGRA* t_cgra, DFG* t_dfg){
+        ifstream refFile("./increMapInput.json");
+        if(!refFile.good()){
+                cout<<"Incremental mapping requires increMapInput.json in current directory!"<<endl;
+                return -1;
+        }
+        json refs;
+        refFile >> refs;
+        CGRANodeID2Level.clear();
```

Comment:
Can you plz add comments about what is `Level`? And what is the relationship between `CGRANodeID` and `Level`

Reply:
In the variable 
map<int, vector<CGRANode*>> FanIO_CGRANodes = {
{4, {CGRANode5, CGRANode6, CGRANode9, CGRANode10}},
{3, {CGRANode1, CGRANode2, CGRANode4, CGRANode7, CGRANode8, CGRANode11, CGRANode13, CGRANode14}},
{2, {CGRANode0, CGRANode3, CGRANode12, CGRANode15}}
}
`Level0` has {CGRANode5, CGRANode6, CGRANode9, CGRANode10}
`Level1` has {CGRANode1, CGRANode2, CGRANode4, CGRANode7, CGRANode8, CGRANode11, CGRANode13, CGRANode14}
`Level2` has {CGRANode0, CGRANode3, CGRANode12, CGRANode15}

Given a CGRANode->getID(), `CGRANodeID2Level` will return its `Level` in the referenced mapping results.